File: Insertion Sort/Insertion Sort.cpp

// Time Complexity: O(n^2)
// Space Complexity: O(1)

#include <iostream>

using namespace std;

void insertion_sort(int *arr, int n)
{
    int j, key;
    for(int i=1; i<n; i++) {
        key = arr[i];
        // Insert A[j] into the sorted sequence A[1..i-1].
        j=i-1;
        while (j>=0 && arr[j]>key)
        {
            arr[j+1] = arr[j];
            j = j-1;
        }
        arr[j+1] = key;
    }
}

void print_array(int *arr, int n) {
    for (int i=0; i<n; i++)
        cout << arr[i] << ' ';
    cout << endl;
}

int main() {
    // Define Array
    int a[] = { 3, 1, 8, 10, 23, 0, -2, 9 };

    // Display unsorted array
    cout << "Unsorted array: ";
    print_array(a, 8);

    // Sort Array In-Place
    insertion_sort(a, 8);

    // Print Sorted Array
    cout << "Sorted array: ";
    print_array(a, 8);

    return 0;
}

File: Insertion Sort/Insertion Sort.py
def insertion_sort(arr):
    n = len(arr)
    for i in range(1, n):
        key = arr[i]
        j = i-1
        while j>=0 and arr[j]>key:
            arr[j+1] = arr[j]
            j = j-1
        arr[j+1] = key

def print_array(arr):
    for element in arr:
        print(element, end=' ')
    print(end='\n')

def main():
    # Define Array
    arr = [3, 1, 8, 10, 23, 0, -2, 9]

    # Display unsorted array
    print("Unsorted array: ", end='')
    print_array(arr)

    insertion_sort(arr)

    print("Sorted  array: ", end='')
    print_array(arr)

if __name__ == '__main__':
    main()
